### DISCUSSION TITLE
Journey through custom setup of SKY130

### DISCUSSION THREAD

**Discussion by suzizecat** [siliconcompiler/siliconcompiler#842] [Help (Q&A)]
Hi there !
I am trying, mainly for understanding purposes, to setup a SKY130 PDK myself.
As I took the existing skywater130.py PDK as an example, I am targeting the same lib : `sky130_fd_sc_hd` 

I installed the PDK through [RTimothyEdwards/open_pdks ](https://github.com/RTimothyEdwards/open_pdks) and modified a copy of skywater130.py to point on the new location. 
The point was to add the full list of buffer cells, clock buffer cells and so on. 

The first issue is that openroad seems to need a single lef file merged containing the standard cell library lef file and the tlef content. 
Could this merging operation be done automatically from within SiliconCompiler flow, as a "read" step of multiple lef files ? 

As of today, specifying the tlef file and the lef file from the PDK, as one (me) would expect, without edition through those lines : 
```python
    # APR tech file
    chip.set('pdk','aprtech',stackup, libtype, 'lef', os.path.join(sclib_dir_ref,"techlef",f"{sclib_name}.tlef"))
    
    # ...
    # lef
    chip.add('library', libname, 'lef', os.path.join(sclib_dir_ref,'lef',f'{sclib_name}.lef'))
```
 produce, at floorplan step, in openroad,
 ```text
[INFO ODB-0222] Reading LEF file: /media/Data/Linux/eda/pdk/sky130A/libs.ref/sky130_fd_sc_hd/techlef/sky130_fd_sc_hd.tlef
[INFO ODB-0223]     Created 13 technology layers
[INFO ODB-0224]     Created 25 technology vias
[INFO ODB-0226] Finished LEF file:  /media/Data/Linux/eda/pdk/sky130A/libs.ref/sky130_fd_sc_hd/techlef/sky130_fd_sc_hd.tlef
[WARNING ODB-0229] Error: library (sky130_fd_sc_hd) already exists
[WARNING ORD-1012] Liberty cell has no LEF master.
[WARNING ORD-1013] instance _393_ LEF master sky130_fd_sc_hd__clkinv_1 not found.
...
```
Followed by a lot of `ORD-1013`, for almost every cell used, see details in [floorplan.log](https://github.com/siliconcompiler/siliconcompiler/files/7853697/floorplan.log)

This might be an issue on OpenROAD side though. 

For reference, find attached my PDK file: [csky130hd.py.txt](https://github.com/siliconcompiler/siliconcompiler/files/7854498/csky130hd.py.txt)  .

**Comment by aolofsson** [comment]
I thought multi-lef issue was fixed a while back in openroad? What version of openroad are you using?
```
openroad -v
````

**Reply by suzizecat** [reply]
It is openroad as installed by siliconcompiler.
`openroad --version` provides `OpenROAD 1 v2.0-880-gd1c7001ad`

**Reply by aolofsson** [reply]
Got it. Looks like the code is a bit out of date. Can you do a git pull. We had to make a change to to the pdk schema to accomodate some corner cases. Looks like your techlef is using the old format? The pypi release 0.4.x and the main repo all follow this new convention. That doesn't explain the issue you are seeing, but it still needs to be done.

chip.set('pdk','aprtech','openroad','M10','12t','lef','tech.lef')

**Reply by suzizecat** [reply]
A git pull of siliconcompiler ? I've done it and it seems that it doesn't fix the issue. 
You can see an updated version of my PDK file on the link in the original post (I've updated it after pulling master version of SC) so now I have : 
```python
    # APR tech file
    for tool in ('openroad', 'klayout', 'magic'):
        chip.set('pdk','aprtech',tool, stackup, libtype, 'lef', os.path.join(sclib_dir_ref,"techlef",f"{sclib_name}.tlef"))
```
for SKY130 which should match the content of skywater130 which is : 
```python
    # Tech file
    for tool in ('openroad', 'klayout', 'magic'):
        chip.set('pdk','aprtech',tool,stackup, libtype,'lef',
                 pdkdir+'/apr/sky130_fd_sc_hd.tlef')
```
I don't really know about techlef format thoug, it comes from SKY130 master too.

**Reply by aolofsson** [reply]
Yes, "a git pull from siliconcompiler." Now that you are up to date, have you done a diff between the tlef that we have included in our repo?
https://github.com/siliconcompiler/siliconcompiler/blob/main/third_party/pdks/skywater/skywater130/pdk/v0_0_2/apr/sky130_fd_sc_hd.tlef

Another possibility, are you running this in an active openroad prompt.
If so you may need to quit out of it to "delete the design", otherwise errors like "library already loaded" would pop up.

Hard to debug like this, so just offering possible some possibilities...

**Reply by suzizecat** [reply]
I'm running it from siliconcompiler only, so no problem on this end.
I'll check the tlef diff, but please know that your help is greatly appreciated nonetheless !

**Reply by suzizecat** [reply]
So, I did the diff and the only difference is that my version got this addition : 
```lef
LAYER nwell
  TYPE MASTERSLICE ;
  PROPERTY LEF58_TYPE "TYPE NWELL ;" ;
END nwell

LAYER pwell
  TYPE MASTERSLICE ;
  PROPERTY LEF58_TYPE "TYPE PWELL ;" ;
END pwell
```
Which should probably not cause any major issue at this point.

**Reply by suzizecat** [reply]
By the way, I get `FOREIGN` statements in the lef file, which seems to cause issues too... Where those removed in siliconcompiler versions ?

**Comment by suzizecat** [comment]
For some reason I get the `pathed_files` variable, in  `def2streazm.py` (lines 140-ish, in the export step) empty when using my PDK while it indeed contains the lef file when I use SC's `skywater130` PDK. Would you have an idea of what could cause this ? 
I indeed provide `chip.add('library', libname, 'lef', os.path.join(sclib_dir_ref,'lef',f'{sclib_name}.merged.lef'))` which should be working, as the synthesis goes fine.

**Reply by suzizecat** [reply]
On this subject, it seems that the `.lyt` file must directly reference the `.lef` file used by the design.
**The lef file passed through silliconcompiler have no impact with regard to this issue.**

Moreover, if the file path is relative, it must start with `./` even if the path following starts with `../../`. See [sky130A_hd.lyt.txt](https://github.com/siliconcompiler/siliconcompiler/files/7866609/sky130A_hd.lyt.txt) for details.

**Comment by nmoroze** [comment]
Thanks for sharing these additional details -- I think one thing that would be helpful is if you could zip/tar your build directory after a run and share that with us? I particularly want to take a look at the manifest files (*.pkg.json) dumped by SC, that would help me debug the problem.

**Reply by suzizecat** [reply]
No problem, I'll make some packages for you.
Just to be sure, is the "trivial" solution (taking the files from the PDK as-is and hoping for the merges operations/libraries path inclusion to KLayout/... to be implicitly done) a good asumption ?

If that's the case, I can make a run with such configuration and give you the build dir, that might be the most meaningful for you.

**Reply by nmoroze** [reply]
Yes, I think that would be the most useful thing to see as a baseline!

**Reply by suzizecat** [reply]
@nmoroze Here is an archive [pdk_test_20220120.tar.gz](https://github.com/siliconcompiler/siliconcompiler/files/7904576/pdk_test_20220120.tar.gz) with:
 - My project `hamster.py` file (no question asked :innocent: ) 
 - Some infos about my PDK setup : (the file tree and SiliconCompiler PDK python file) in `pdk/`
 - The full build dir 
 
Let me know if you need more information.

**Reply by nmoroze** [reply]
Thank you! I think it would also be helpful to see copies of the following files:

- sky130_fd_sc_hd.lef
- sky130_fd_sc_hd.merged.cleaned.lef
 - sky130_fd_sc_hd.merged.lef
 - sky130_fd_sc_hd.tlef

My best guess right now is that this is something subtle to do with differences in LEF content. I don't believe OpenROAD needs the tlef and library data merged, rather I think the "merged" in the filename in our case refers to a merged file containing all std cell definitions. We took our LEF files from the [OpenROAD-flow-scripts repo](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/tree/master/flow/platforms/sky130hd/lef), rather than directly from the open PDKs repo (so we were not responsible for the preprocessing).

**Reply by suzizecat** [reply]
I'll send them to you, just for information, merged is the merging of the .tlef + the .lef using mergeLef tool from the openroad utils. The cleaned was an unsuccessful attempt and is not meaningful.

**Reply by suzizecat** [reply]
Please find all mentioned files in this archive : [sky130_fd_sc_hd_lef.tar.gz](https://github.com/siliconcompiler/siliconcompiler/files/7911488/sky130_fd_sc_hd_lef.tar.gz)

**Reply by aolofsson** [reply]
@suzizecat I am hoping this is a dead thread after 3 years, but just checking. A lot has happened with SC since 2022.

**Reply by suzizecat** [reply]
Hi @aolofsson , It is indeed a dead thread (mainly because it was an experiment more than anything, so I went on other experiments) but IIRC I never quite fixed those issues in the end. 

I'll give another spin to SC when I'll have a suitable project to play around with 🙂
